homeassistant/util/yaml.py

```diff
@@ -162,18 +160,22 @@ def _secret_yaml(loader: SafeLineLoader,
                                   " but 'logger: %s' found", logger)
                 del secrets['logger']
         else:
-            loader._SECRET_CACHE[secret_path] = None
-    secrets = loader._SECRET_CACHE[secret_path]
+            _SECRET_CACHE[secret_path] = {}
+    secrets = _SECRET_CACHE[secret_path]
 
     # Retrieve secret, first from secrets.yaml, then from keyring
     if secrets is not None and node.value in secrets:
         _LOGGER.debug('Secret %s retrieved from secrets.yaml.', node.value)
         return secrets[node.value]
-    for sname, sdict in loader._SECRET_CACHE.items():
-        if node.value in sdict:
+
+    while os.path.exists(secret_path) and secret_path in _SECRET_CACHE:
+        _LOGGER.debug('Checking path %s', secret_path)
+        secrets = _SECRET_CACHE[secret_path]
+        if node.value in secrets:
             _LOGGER.debug('Secret %s retrieved from secrets.yaml in other '
-                          'folder %s', node.value, sname)
-            return sdict[node.value]
+                          'folder %s', node.value, secret_path)
+            return secrets[node.value]
+        secret_path = os.path.abspath(os.path.join(secret_path, '..'))
```

Comment:
I'm not super familiar with pythong, but I thought `dirname()` returned just the directory name of the file, i.e. `/home/user/.homeassistant/automation/myauto.yaml` -> `automation`